### PR TITLE
Kernel cleanup

### LINKED ISSUE #44: Wrong check in kernel name validation
The `KernelCore` constructor [checks](https://github.com/nuun-io/kernel/blob/e9388d38cfdbcf3c1f20c18aa7de9ddd0d554ec8/core/src/main/java/io/nuun/kernel/core/internal/KernelCore.java#L109) that the kernel name is unique. The actual check doesn't work, it should be replaced by `kernels.containsKey(name())`.

```
// private static final ConcurrentHashMap<String, Kernel> kernels  = new ConcurrentHashMap<String, Kernel>();

if (!kernels.contains(name()))
{
    kernels.put(name(), this);
}
else
{
    throw new KernelException(String.format("A kernel named %s already exists", name()));
}
```

Otherwise, it seems that this map is only used to check the name uniqueness. So is it really necessary to store kernel instances in the map ?

Currently, the name cannot be configured. So maybe, if we could have a safer name initialization, the map could be removed. For instance at this [line](https://github.com/nuun-io/kernel/blob/e9388d38cfdbcf3c1f20c18aa7de9ddd0d554ec8/core/src/main/java/io/nuun/kernel/core/internal/KernelCore.java#L105), we could use an `AtomicInteger`.

    name = KERNEL_PREFIX_NAME + kernelIndex.incrementAndGet();

### PR DESCRIPTION
**Pull request by pith**
- KernelCore refactoring
- Improve some javadoc
- Fix #44 kernel name
- Improve logging
- Start a new IT package

### CODE REVIEW

File: core/src/main/java/io/nuun/kernel/core/internal/KernelConfigurationInternal.java [hunk 1]
```diff
@@ -92,16 +82,28 @@ public KernelConfiguration containerContext(Object containerContext)
...
-        this.pluginsClass = concat(this.pluginsClass, pluginsClass);
+        this.plugins = ObjectArrays.concat(plugin, this.plugins);
         return this;
     }
 
     @Override
     public KernelConfiguration plugins(Plugin... plugins)
     {
-        this.plugins = this.<Plugin>concat(this.plugins, plugins);
+        this.plugins = ObjectArrays.concat(this.plugins, plugins, Plugin.class);
```
Comment (ejemba) [inline core/src/main/java/io/nuun/kernel/core/internal/KernelConfigurationInternal.java]: `, Plugin.class` here I do not understand why we need this.

Comment (pith) [inline core/src/main/java/io/nuun/kernel/core/internal/KernelConfigurationInternal.java]: It's required by the `concat()` method in the case where the two params are arrays (cf. http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/ObjectArrays.html)

File: core/src/main/java/io/nuun/kernel/core/internal/KernelCore.java [hunk 2]
```diff
@@ -159,15 +144,22 @@ public synchronized void init()
         }
     }
 
-    private void initRoundEnvironment()
+    public void preparePlugins(List<Plugin> fetchedPlugins) {
+        initRoundEnvironment(fetchedPlugins);
+        checkPlugins(fetchedPlugins);
+        computeAliases(fetchedPlugins);
+        fetchGlobalParametersFromPlugins();
```
Comment (ejemba) [inline core/src/main/java/io/nuun/kernel/core/internal/KernelCore.java]: According to your previous changes, don't `fetchGlobalParametersFromPlugins()` could take `fetchedPlugins` as parameters as well ?

File: specs/src/main/java/io/nuun/kernel/spi/DependencyInjectionProvider.java [hunk 3]
```diff
@@ -18,26 +18,38 @@
...
+     * </p>
+     *
+     * @param unitModule an object containing binding definitions
      *
-     * @param injectionDefinition
-     * @return Guice Plugin
+     * @return a Nuun unit module
      */
-    UnitModule convert(Object injectionDefinition);
+    UnitModule convert(Object unitModule);
```
Comment (ejemba) [inline specs/src/main/java/io/nuun/kernel/spi/DependencyInjectionProvider.java]: The parameter given is technically not a unit module but something else (to be converted to a unitModule). 
Perhaps naming the parameter `unitModule` can be confusing, I think.

Comment (pith) [inline specs/src/main/java/io/nuun/kernel/spi/DependencyInjectionProvider.java]: Yes I have to admit I'm sometimes lost with all the injection concepts. Maybe `BindingDefinitions` would have been a better name ?

Comment (ejemba) [inline specs/src/main/java/io/nuun/kernel/spi/DependencyInjectionProvider.java]: Yes  `UnitModule convert(Object bindingDefinitions);` sounds very good. 

PS : I admit without a bigpicture of the concepts it is not that easy. I'm working on it.

**Comment by pith** [comment]
@ejemba I changed variable name like asked. Tell me if its ok for you.
I have another PR pending, but this one should be merged before.

**Pull request merged**